File: scripts/manage_users.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "bcrypt",
#     "typer",
#     "iterfzf",
# ]
# ///
"""Manage users in the SQLite auth database.

Reads DATABASE_PATH from the environment (loaded by justfile's dotenv-load).
"""

import json
import os
import sqlite3
import subprocess
import uuid

import bcrypt
import typer
from iterfzf import iterfzf

app = typer.Typer()
DEFAULT_ROLE = "viewer"


def get_db_path() -> str:
    path = os.environ.get("DATABASE_PATH")
    if not path:
        typer.echo("Error: DATABASE_PATH environment variable is not set.")
        raise typer.Exit(1)
    return path


def ensure_schema():
    """Run the node init-db script to create tables if they don't exist."""
    script = os.path.join(os.path.dirname(__file__), "init-db.ts")
    subprocess.run(["node", "--experimental-strip-types", script], check=True)


def get_db_connection() -> sqlite3.Connection:
    conn = sqlite3.connect(get_db_path())
    conn.execute("PRAGMA journal_mode = WAL")
    conn.execute("PRAGMA foreign_keys = ON")
    return conn


def get_users() -> list[str]:
    conn = get_db_connection()
    cursor = conn.execute("SELECT username FROM user_credentials")
    users = [row[0] for row in cursor.fetchall()]
    conn.close()
    return users


def generate_password() -> str:
    result = subprocess.run(
        ["openssl", "rand", "-base64", "32"],
        capture_output=True,
        text=True,
        check=True,
    )
    return result.stdout.strip()


def log_audit_event(cursor, user_id, username, event_type, details="", meta=None):
    """Insert a row into audit_log within the caller's transaction."""
    cursor.execute(
        """INSERT INTO audit_log (user_id, username, event_type, details, meta)
           VALUES (?, ?, ?, ?, ?)""",
        (user_id, username, event_type, details, json.dumps(meta) if meta else None),
    )


def hash_password(password: str) -> str:
    salt = bcrypt.gensalt(rounds=12)
    hashed = bcrypt.hashpw(password.encode("utf-8"), salt)
    return hashed.decode("utf-8")


@app.command()
def create_user(username: str = typer.Argument(..., help="Username for the new user")):
    """Create a new user in the user_credentials table."""

    conn = get_db_connection()
    row = conn.execute(
        "SELECT username FROM user_credentials WHERE username = ?", (username,)
    ).fetchone()
    if row:
        typer.echo(f"Error: Username '{username}' already exists!")
        conn.close()
        raise typer.Exit(1)
    conn.close()

    user_id = str(uuid.uuid4())
    password = generate_password()
    password_hash = hash_password(password)

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute(
            """INSERT INTO user_credentials (user_id, username, password_hash, role)
               VALUES (?, ?, ?, ?)""",
            (user_id, username, password_hash, DEFAULT_ROLE),
        )
        log_audit_event(
            cursor, user_id, username, "user_created",
            f"Role: {DEFAULT_ROLE}", {"source": "cli"},
        )
        conn.commit()

        typer.echo(f"User created successfully!")
        typer.echo(f"User ID: {user_id}")
        typer.echo(f"Username: {username}")
        typer.echo(f"Generated Password: {password}")
        typer.echo(f"Save this password - it cannot be retrieved later!")

    except sqlite3.Error as e:
        typer.echo(f"Error creating user: {e}")
        raise typer.Exit(1)
    finally:
        conn.close()


@app.command()
def update_password():
    """Update the password for an existing user."""

    users = get_users()
    if not users:
        typer.echo("No users found in the database!")
        raise typer.Exit(1)

    selected_username = iterfzf(users, prompt="Select user to update password: ")

    if not selected_username:
        typer.echo("No user selected. Aborting.")
        raise typer.Exit(1)

    new_password = generate_password()
    password_hash = hash_password(new_password)

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute(
            "SELECT user_id FROM user_credentials WHERE username = ?",
            (selected_username,),
        )
        row = cursor.fetchone()

        if row is None:
            typer.echo(f"Error: User '{selected_username}' not found!")
            raise typer.Exit(1)

        user_id = row[0]
        cursor.execute(
            "UPDATE user_credentials SET password_hash = ? WHERE username = ?",
            (password_hash, selected_username),
        )
        log_audit_event(
            cursor, user_id, selected_username, "password_reset",
            "Password reset via CLI", {"source": "cli"},
        )
        conn.commit()
        typer.echo(f"Password updated successfully for user: {selected_username}")
        typer.echo(f"New Password: {new_password}")
        typer.echo(f"Save this password - it cannot be retrieved later!")

    except sqlite3.Error as e:
        typer.echo(f"Error updating password: {e}")
        raise typer.Exit(1)
    finally:
        conn.close()


@app.command()
def delete_user():
    """Delete an existing user from user_credentials."""

    users = get_users()
    if not users:
        typer.echo("No users found in the database!")
        raise typer.Exit(1)

    selected_username = iterfzf(users, prompt="Select user to delete: ")

    if not selected_username:
        typer.echo("No user selected. Aborting.")
        raise typer.Exit(1)

    confirm = typer.confirm(f"Are you sure you want to delete user '{selected_username}'?")
    if not confirm:
        typer.echo("Deletion cancelled.")
        return

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute(
            "SELECT user_id FROM user_credentials WHERE username = ?",
            (selected_username,),
        )
        row = cursor.fetchone()

        if row is None:
            typer.echo(f"Error: User '{selected_username}' not found!")
            raise typer.Exit(1)

        user_id = row[0]
        cursor.execute(
            "DELETE FROM user_credentials WHERE username = ?",
            (selected_username,),
        )
        log_audit_event(
            cursor, user_id, selected_username, "user_deleted",
            "User deleted via CLI", {"source": "cli"},
        )
        conn.commit()
        typer.echo(f"User '{selected_username}' deleted successfully.")

    except sqlite3.Error as e:
        typer.echo(f"Error deleting user: {e}")
        raise typer.Exit(1)
    finally:
        conn.close()


if __name__ == "__main__":
    ensure_schema()
    app()

File: scripts/manage_users.ts
#!/usr/bin/env bun
/**
 * Manage users in the SQLite auth database.
 *
 * Reads DATABASE_PATH from the environment (loaded by justfile's dotenv-load).
 *
 * Usage:
 *   bun run scripts/manage_users.ts create-user <username>
 *   bun run scripts/manage_users.ts update-password
 *   bun run scripts/manage_users.ts delete-user
 */
import { Database } from "bun:sqlite";
import { confirm, select } from "@inquirer/prompts";
import crypto from "node:crypto";
import { hashPassword } from "~/lib/auth/hash";

const DEFAULT_ROLE = "viewer";

function getDb(): Database {
  const dbPath = process.env.DATABASE_PATH;
  if (!dbPath) {
    console.error("Error: DATABASE_PATH environment variable is not set.");
    process.exit(1);
  }
  const db = new Database(dbPath);
  db.run("PRAGMA journal_mode = WAL");
  db.run("PRAGMA foreign_keys = ON");
  return db;
}

const db = getDb();

function generatePassword(): string {
  return crypto.randomBytes(32).toString("base64url");
}

function listUsernames(): string[] {
  return db
    .query("SELECT username FROM user_credentials")
    .all()
    .map((r) => (r as { username: string }).username);
}

function logAuditEvent(
  userId: string | null,
  username: string,
  eventType: string,
  details: string,
  meta?: Record<string, unknown>,
): void {
  db.run(
    "INSERT INTO audit_log (user_id, username, event_type, details, meta) VALUES (?, ?, ?, ?, ?)",
    [userId, username, eventType, details, meta ? JSON.stringify(meta) : null],
  );
}

async function createUser(username: string): Promise<void> {
  const existing = db
    .query("SELECT 1 FROM user_credentials WHERE username = ?")
    .get(username);
  if (existing) {
    console.error(`Error: Username '${username}' already exists!`);
    process.exit(1);
  }

  const userId = crypto.randomUUID();
  const password = generatePassword();
  const passwordHash = await hashPassword(password);

  db.run(
    "INSERT INTO user_credentials (user_id, username, password_hash, role) VALUES (?, ?, ?, ?)",
    [userId, username, passwordHash, DEFAULT_ROLE],
  );

  logAuditEvent(userId, username, "user_created", `Role: ${DEFAULT_ROLE}`, {
    source: "cli",
  });

  console.log("User created successfully!");
  console.log(`User ID: ${userId}`);
  console.log(`Username: ${username}`);
  console.log(`Generated Password: ${password}`);
  console.log("Save this password - it cannot be retrieved later!");
}

async function updatePassword(): Promise<void> {
  const users = listUsernames();
  if (users.length === 0) {
    console.log("No users found in the database!");
    process.exit(1);
  }

  const selectedUsername = await select({
    message: "Select user to update password:",
    choices: users.map((u) => ({ name: u, value: u })),
  });

  const row = db
    .query("SELECT user_id FROM user_credentials WHERE username = ?")
    .get(selectedUsername) as { user_id: string } | null;
  if (!row) {
    console.error(`Error: User '${selectedUsername}' not found!`);
    process.exit(1);
  }

  const newPassword = generatePassword();
  const passwordHash = await hashPassword(newPassword);

  db.run("UPDATE user_credentials SET password_hash = ? WHERE username = ?", [
    passwordHash,
    selectedUsername,
  ]);

  logAuditEvent(
    row.user_id,
    selectedUsername,
    "password_reset",
    "Password reset via CLI",
    { source: "cli" },
  );

  console.log(`Password updated successfully for user: ${selectedUsername}`);
  console.log(`New Password: ${newPassword}`);
  console.log("Save this password - it cannot be retrieved later!");
}

async function deleteUser(): Promise<void> {
  const users = listUsernames();
  if (users.length === 0) {
    console.log("No users found in the database!");
    process.exit(1);
  }

  const selectedUsername = await select({
    message: "Select user to delete:",
    choices: users.map((u) => ({ name: u, value: u })),
  });

  const confirmed = await confirm({
    message: `Are you sure you want to delete user '${selectedUsername}'?`,
    default: false,
  });

  if (!confirmed) {
    console.log("Deletion cancelled.");
    return;
  }

  const row = db
    .query("SELECT user_id FROM user_credentials WHERE username = ?")
    .get(selectedUsername) as { user_id: string } | null;
  if (!row) {
    console.error(`Error: User '${selectedUsername}' not found!`);
    process.exit(1);
  }

  db.run("DELETE FROM user_credentials WHERE username = ?", [selectedUsername]);

  logAuditEvent(
    row.user_id,
    selectedUsername,
    "user_deleted",
    "User deleted via CLI",
    { source: "cli" },
  );

  console.log(`User '${selectedUsername}' deleted successfully.`);
}

// --- Main ---

const [command, ...args] = process.argv.slice(2);

switch (command) {
  case "create-user": {
    const username = args[0];
    if (!username) {
      console.error("Usage: manage_users.ts create-user <username>");
      process.exit(1);
    }
    await createUser(username);
    break;
  }
  case "update-password":
    await updatePassword();
    break;
  case "delete-user":
    await deleteUser();
    break;
  default:
    console.log("Usage: manage_users.ts <command>");
    console.log("");
    console.log("Commands:");
    console.log("  create-user <username>  Create a new user");
    console.log("  update-password         Update password for an existing user");
    console.log("  delete-user             Delete an existing user");
    process.exit(command ? 1 : 0);
}
